Code/GraphMol/MolStandardize/Normalize.cpp

```diff
@@ -98,30 +97,31 @@ ROMol *Normalizer::normalize(const ROMol &mol) {
   return outmol;
 }
 
-boost::shared_ptr<ROMol> Normalizer::normalizeFragment(
+ROMOL_SPTR Normalizer::normalizeFragment(
     const ROMol &mol,
-    const std::vector<std::shared_ptr<ChemicalReaction>> &transforms) {
-  boost::shared_ptr<ROMol> nfrag(new ROMol(mol));
+    const std::vector<std::shared_ptr<ChemicalReaction>> &transforms) const {
+  ROMOL_SPTR nfrag(new ROMol(mol));
   MolOps::fastFindRings(
       *nfrag);  // this doesn't do anything if rings are already there
+  std::set<std::string> seenProductSmiles;
   for (unsigned int i = 0; i < MAX_RESTARTS; ++i) {
-    bool loop_brake = false;
+    bool loop_break = false;
     // Iterate through Normalization transforms and apply each in order
     for (auto &transform : transforms) {
-      boost::shared_ptr<ROMol> product =
-          this->applyTransform(nfrag, *transform);
-      if (product != nullptr) {
+      SmilesMolPair product = applyTransform(nfrag, *transform);
+      if (!product.first.empty() and !seenProductSmiles.count(product.first)) {
+        seenProductSmiles.insert(product.first);
         BOOST_LOG(rdInfoLog)
             << "Rule applied: "
             << transform->getProp<std::string>(common_properties::_Name)
             << "\n";
-        nfrag = product;
-        loop_brake = true;
+        nfrag = product.second;
+        loop_break = true;
         break;
       }
     }
     // For loop finishes normally, all applicable transforms have been applied
-    if (!loop_brake) {
+    if (!loop_break) {
```

Comment:
Using `loop_break` outside the scope where it was defined.

Reply:
It is not out of scope, I think MSVC got deeply confused by the `and`. I really only fixed a typo there.